src/Microsoft.Diagnostics.ExtensionCommands/DumpHeapCommand.cs

```diff
@@ -0,0 +1,381 @@
+                if (!char.IsLetterOrDigit(ch))
+                {
+                    return FilterString(str, maxLen);
+                }
+            }
+
+            return str;
+
+            static string FilterString(string str, int maxLen)
+            {
+                maxLen = Math.Min(str.Length, maxLen);
+                Debug.Assert(maxLen <= 128);
+
+                Span<char> buffer = stackalloc char[maxLen];
+                ReadOnlySpan<char> value = str.AsSpan(0, buffer.Length);
+
+                for (int i = 0; i < value.Length; ++i)
+                {
+                    char ch = value[i];
+                    buffer[i] = char.IsLetterOrDigit(ch) || char.IsPunctuation(ch) || ch == ' ' ? ch : StringReplacementCharacter;
+                }
+
+                return buffer.ToString();
+            }
+        }
+
+        private void ParseArguments()
+        {
+            if (Live && Dead)
+            {
+                Live = false;
+                Dead = false;
+            }
+
+            if (!string.IsNullOrWhiteSpace(MethodTableString))
+            {
+                if (ParseHexString(MethodTableString, out ulong mt))
+                {
+                    MethodTable = mt;
+                }
+                else
+                {
+                    throw new ArgumentException($"Invalid MethodTable: {MethodTableString}");
+                }
+            }
+
+            FilteredHeap = new(Runtime.Heap);
+            if (GCHeap >= 0)
+            {
+                FilteredHeap.GCHeap = GCHeap;
+            }
+
+            if (!string.IsNullOrWhiteSpace(Segment))
+            {
+                FilteredHeap.FilterBySegmentHex(Segment);
+            }
+
+            if (MemoryRange is not null && MemoryRange.Length > 0)
+            {
+                if (MemoryRange.Length > 2)
+                {
+                    string badArgument = MemoryRange.FirstOrDefault(f => f.StartsWith("-") || f.StartsWith("/"));
+                    if (badArgument != null)
+                    {
+                        throw new ArgumentException($"Unknown argument: {badArgument}");
+                    }
+
+                    throw new ArgumentException("Too many arguments to !dumpheap");
+                }
+
+                string start = MemoryRange[0];
+                string end = MemoryRange.Length > 1 ? MemoryRange[1] : null;
+                FilteredHeap.FilterByHexMemoryRange(start, end);
+            }
+
+            if (Min > 0)
+            {
+                FilteredHeap.MinimumObjectSize = Min;
+            }
+
+            if (Max > 0)
+            {
+                FilteredHeap.MaximumObjectSize = Max;
+            }
+
+            if (Strings)
+            {
+                MethodTable = Runtime.Heap.StringType.MethodTable;
+            }
+
+            FilteredHeap.SortSegments = (seg) => seg.OrderBy(seg => seg.Start);
+        }
+
+        private static bool ParseHexString(string str, out ulong value)
```

Comment:
The CommandBase has a TryParseAddress function that does the same thing.  If the CommandBase has limitations or problems that your implementation fixes could you move them to CommandBase.TryParseAddress and use that one.  

Someday I plan to add a expression evaluation service (it will call the underlying host debugger's expression evaluator) that will hook into the command service parser that makes all of this moot. Using this CommandBase function will help me find all the places the manual parsing is done.

Reply:
Thanks, I missed that.  I'll make that change.

It would be really nice if that was part of the CommandAttribute.  I want to write this:

``` C#
[Option(Name="Foo", Format="HexAddress")]
public ulong Address { get; set }
```

But as I was poking around System.CommandLine it wasn't obvious how set a custom parser like that, so I've left it for another day.

Reply:
There is a way to hook SystemCommandLine option/argument parsing theoretically. I haven't tried it yet but it is on my list as part of an official IExpressionService.  And yes, adding hint properties to the option attr is a good idea.

Reply:
I'm going to tackle this on my next PR.  I want to get this merged so I can rebase my other PR without too much drift.